### PR TITLE
[WIP] [image_picker_for_web] Added support for maxWidth, maxHeight and imageQuality 

### PR DESCRIPTION
**Pull request by balvinderz**
Adds support for maxWidth, maxHeight and imageQuality in image_picker_for_web

https://github.com/flutter/flutter/issues/78170

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Note that unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [X] I signed the [CLA].
- [X] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I [updated pubspec.yaml](https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates) with an appropriate new version according to the [pub versioning philosophy].
- [X] I updated CHANGELOG.md to add a description of the change.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code

### CODE REVIEW

**Comment by balvinderz** [comment]
Implementation is based on this : https://labs.madisoft.it/javascript-image-compression-and-resizing/

There are some limitations in this.
imageQuality only works for jpeg images.
I could not find a way to make this work for gif images.

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart [hunk 1]
```diff
@@ -121,6 +121,9 @@ class ImagePickerPlugin extends ImagePickerPlatform {
     List<XFile> files = await getFiles(
       accept: _kAcceptImageMimeType,
       capture: capture,
+      maxWidth: maxWidth,
+      maxHeight: maxHeight,
+      imageQuality: imageQuality,
```
Comment (ditman) [inline packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:126]: Please, do not pass the image quality parameters to the getFiles method, otherwise you have to drill those down through several layers of the plugin.

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart [hunk 2]
```diff
@@ -121,6 +121,9 @@ class ImagePickerPlugin extends ImagePickerPlatform {
...
     );
     return files.first;
```
Comment (ditman) [inline packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:128]: Instead, determine if resizing is needed, and then `return imageResizer.resizeImage(files.first, maxWidth, maxHeight, imageQuality);`

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart [hunk 3]
```diff
@@ -158,7 +161,12 @@ class ImagePickerPlugin extends ImagePickerPlatform {
     double? maxHeight,
     int? imageQuality,
   }) {
-    return getFiles(accept: _kAcceptImageMimeType, multiple: true);
```
Comment (ditman) [inline packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:161]: Same here, determine if resizing is needed, and then return imageResizer.resizeImages(getFiles(...), maxWidth, maxHeight, imageQuality);

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart [hunk 4]
```diff
@@ -171,19 +179,22 @@ class ImagePickerPlugin extends ImagePickerPlatform {
...
-    String? capture,
-    bool multiple = false,
-  }) {
+  Future<List<XFile>> getFiles(
+      {String? accept,
+      String? capture,
+      bool multiple = false,
+      double? maxWidth,
+      double? maxHeight,
+      int? imageQuality}) {
```
Comment (ditman) [inline packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:188]: (Do not accept the resize image parameters any further down than the getImage/getImages method)

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart [hunk 5]
```diff
@@ -238,23 +249,29 @@ class ImagePickerPlugin extends ImagePickerPlatform {
...
-                  lastModified: DateTime.fromMillisecondsSinceEpoch(
-                    file.lastModified ?? DateTime.now().millisecondsSinceEpoch,
-                  ),
-                  mimeType: file.type,
-                ))
-            .toList());
+        //Convert the file  if one of the parameter is not null
+        if (imageQuality != null || maxWidth != null || maxHeight != null) {
+          final convertedFileUris = await Future.wait(files.map((e) async =>
+              await _getCompressedUri(e, imageQuality, maxHeight, maxWidth)));
```
Comment (ditman) [inline packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:262]: This `_getCompressedUri` method should be a method from a separate object that only deals with resizing images in `XFile` objects.

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart [hunk 6]
```diff
@@ -310,6 +327,73 @@ class ImagePickerPlugin extends ImagePickerPlatform {
     _target.children.add(element);
     element.click();
   }
+
+  Future<String> _getCompressedUri(html.File file, int? imageQuality,
```
Comment (ditman) [inline packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:331]: Extract to a new `ImageResizer` object. Also, try to use: Future<XFile> as the input format for the file, and the output for the resized image.

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart [hunk 7]
```diff
@@ -310,6 +327,73 @@ class ImagePickerPlugin extends ImagePickerPlatform {
...
+    });
+    image.onError.listen((event) {
+      //Return the original image if error comes
+      _completer.complete(html.Url.createObjectUrl(file));
+    });
+    return _completer.future;
+  }
+
+  List<int> _calculateSize(
+      html.ImageElement img, double maxWidth, double maxHeight) {
```
Comment (ditman) [inline packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:367]: Move this helper to the `ImageResizer` object as well. Also, maybe return a `Size` object [docs](https://api.flutter.dev/flutter/dart-ui/Size-class.html), instead of a list of ints?

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart [hunk 8]
```diff
@@ -310,6 +327,73 @@ class ImagePickerPlugin extends ImagePickerPlatform {
...
+extension on html.File {
+  XFile toXFile({String? path}) => XFile(
+        path ?? html.Url.createObjectUrl(this),
+        name: this.name,
+        length: this.size,
+        lastModified: DateTime.fromMillisecondsSinceEpoch(
+          this.lastModified ?? DateTime.now().millisecondsSinceEpoch,
+        ),
+        mimeType: this.type,
+      );
```
Comment (ditman) [inline packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:396]: Let's not apply this extension on a dart:html class. This will cause people to some times have a `.toXFile` method on their `html.File`, and some times **not** have it, and the cause is whether they use the image_picker_for_web or not.

(This might be more appropriate for the cross_file package, but I think the best solution is still a XFile constructor for web only that takes an html.File input file.)

**Review by ditman — CHANGES_REQUESTED** [review]
Thanks for the contribution @balvinderz! I think that this one is a little bit too green, and it needs some changes to the approach to make this more modular and testable!

In my opinion, the image resizing code should be extracted from the `ImagePickerPlugin` class completely. The only thing the plugin knows about resizing images is that it has an instance of an `ImageResizer` object that accepts and returns `XFile` instances.

I think the following steps are needed:

- [ ] Create a `ImageResizer` component that deals with images in `XFile` instances, and resizes them according to the parameters passed by the user.
  - 99% of the code of this component already exists in the `_getCompressedUri` and `_calculateSize` methods. Having your own object will allow you to design it more carefully. 
- [ ] Have an instance of that `ImageResizer` inside of the `ImagePickerPlugin`, so it can be:
  - Used when needed to resize images (normal operation)
  - Injectable from tests so it can be mocked out if needed (testability)
- [ ] Use the `ImageResizer` in the `pickImage`/`pickImages` methods, if the user wants a resized output
  - **Do not pass the image resizing parameters (`maxWidth`, `maxHeight`, `imageQuality`) to the generic methods that pick files and convert them to a XFile!**
- [ ] Unit tests for the new `ImageResizer` component
- [ ] Unit tests for the new resize images (when the correct parameters are passed) feature.

**Comment by balvinderz** [comment]
I think i have 

> Thanks for the contribution @balvinderz! I think that this one is a little bit too green, and it needs some changes to the approach to make this more modular and testable!
> 
> In my opinion, the image resizing code should be extracted from the `ImagePickerPlugin` class completely. The only thing the plugin knows about resizing images is that it has an instance of an `ImageResizer` object that accepts and returns `XFile` instances.
> 
> I think the following steps are needed:
> 
> * [X]  Create a `ImageResizer` component that deals with images in `XFile` instances, and resizes them according to the parameters passed by the user.
>   
>   * 99% of the code of this component already exists in the `_getCompressedUri` and `_calculateSize` methods. Having your own object will allow you to design it more carefully.
> * [X]  Have an instance of that `ImageResizer` inside of the `ImagePickerPlugin`, so it can be:
>   
>   * Used when needed to resize images (normal operation)
>   * Injectable from tests so it can be mocked out if needed (testability)
> *  [X]  Use the `ImageResizer` in the `pickImage`/`pickImages` methods, if the user wants a resized output
>   
>   * **Do not pass the image resizing parameters (`maxWidth`, `maxHeight`, `imageQuality`) to the generic methods that pick files and convert them to a XFile!**
> * [X]  Unit tests for the new `ImageResizer` component
> * [X]  Unit tests for the new resize images (when the correct parameters are passed) feature.

I think I have done all the steps that are mentioned in this.

**Comment by ditman** [comment]
> I think I have done all the steps that are mentioned in this.

Amazing, thanks! I'll take a look later tonight!

**Comment by balvinderz** [comment]
I have found a way to scale gif images using package:image. 
https://pub.dev/packages/image
But it is way too slow and freezes the ui.

File: packages/image_picker/image_picker_for_web/lib/image_resizer_utils.dart [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+bool isImageQualityValid(int imageQuality) {
+  return (imageQuality >= 0 && imageQuality <= 100);
+}
+
+/// a functions that calculates the size of the scaled image.
+/// imageWidth is the width of the image
+/// imageHeight is the height of  the image
+/// maxWidth is the maximum width of the scaled image
+/// maxHeight is the maximum height of the scaled image
+Size calculateSizeOfScaledImage(double imageWidth, double imageHeight,
```
Comment (ditman) [inline packages/image_picker/image_picker_for_web/lib/image_resizer_utils.dart:33]: If you pass `Size imageSize` in, you can use the tools in the Size object to do the resizing (see [operator *](https://api.flutter.dev/flutter/dart-ui/Size/operator_multiply.html))

File: packages/image_picker/image_picker_for_web/lib/image_resizer_utils.dart [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+    double? maxWidth, double? maxHeight) {
```
Comment (ditman) [inline packages/image_picker/image_picker_for_web/lib/image_resizer_utils.dart:34]: What is the behavior here when both maxWidth and maxHeight are passed? Constrain to fit? Stretch to fit?

Comment (ditman) [reply packages/image_picker/image_picker_for_web/lib/image_resizer_utils.dart:34]: If we never stretch, but constrain to the smallest side maitaining aspect ratio, the code can be simplified a lot, I think!

File: packages/image_picker/image_picker_for_web/lib/image_resizer_utils.dart [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+  double originalWidth = imageWidth;
+  double originalHeight = imageHeight;
+
+  bool hasMaxWidth = maxWidth != null;
```
Comment (ditman) [inline packages/image_picker/image_picker_for_web/lib/image_resizer_utils.dart:38]: If we're not stretching the output of the image, I think most of this method can be simplified a little bit:

```dart
Size calculateSizeOfScaledImage(Size imageSize, double? maxWidth, double? maxHeight) {
  double widthFactor = maxWidth != null ? imageSize.width / maxWidth : 0;
  double heightFactor = maxHeight != null ? imageSize.height / maxHeight : 0;
  double resizeFactor = Math.max(widthFactor, heightFactor);
  return imageSize ~/ resizeFactor;
}
```

**Review by ditman — COMMENTED** [review]
I have some general comments. I think the image_resizer_utils can be revisited with much simpler algorithms (note that it's only Github code, so I haven't tested that my suggestions have 100% perfect syntax or cover all the corner cases that your code might do!), please take them as inspiration to cut down the amount of LOC, if possible! ;)

File: packages/image_picker/image_picker_for_web/lib/image_resizer_utils.dart [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+  if (resizeNeeded) {
+    if (imageQuality != null) {
+      return isImageQualityValid(imageQuality);
+    } else {
+      return true;
+    }
+  } else {
+    return false;
+  }
+}
```
[the remark this answers is not in the document]
Comment (balvinderz) [reply packages/image_picker/image_picker_for_web/lib/image_resizer_utils.dart:21]: yes

File: packages/image_picker/image_picker_for_web/lib/image_resizer_utils.dart (hunk 10 above)
Comment (balvinderz) [reply packages/image_picker/image_picker_for_web/lib/image_resizer_utils.dart:34]: We are never stretching. My implementation of this function was based on how  the ios side of this plugin scales the image.

File: packages/image_picker/image_picker_for_web/lib/image_resizer_utils.dart (hunk 11 above)
Comment (balvinderz) [reply packages/image_picker/image_picker_for_web/lib/image_resizer_utils.dart:38]: This does work but it scales the image above its original width and height . I think I should have used the word downScaled instead of scaled in this function .  The android and IOS implementation of this function always downScales the  image and never upscales the image.  so I think, I will keep the same implementation as the others to be consistent in all platforms . What do you think?

File: packages/image_picker/image_picker_for_web/lib/image_resizer.dart [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+import 'dart:math';
+import 'package:image_picker_for_web/image_resizer_utils.dart';
+import 'package:image_picker_platform_interface/image_picker_platform_interface.dart';
+import 'dart:html' as html;
+
+/// Helper class that resizes image
+class ImageResizer {
+  /// Resizes the image if needed
+  /// Does not support gif image
+  Future<XFile> resizeImageIfNeeded(XFile file, double? maxWidth,
```
[the remark this answers is not in the document]
Comment (balvinderz) [reply packages/image_picker/image_picker_for_web/lib/image_resizer.dart:16]: I have now  made three functions . Would they need to be individually tested ?.

File: packages/image_picker/image_picker_for_web/lib/image_resizer_utils.dart (hunk 11 above)
Comment (ditman) [reply packages/image_picker/image_picker_for_web/lib/image_resizer_utils.dart:30]: > The android and IOS implementation of this function always downScales the image and never upscales the image.

Absolutely, yes!

File: packages/image_picker/image_picker_for_web/README.md [hunk 14]
```diff
@@ -43,7 +43,8 @@ Each browser may implement `capture` any way they please, so it may (or may not)
 difference in your users' experience.
 
 ### pickImage()
-The arguments `maxWidth`, `maxHeight` and `imageQuality` are not supported on the web.
+The arguments `maxWidth`, `maxHeight` and `imageQuality` are not supported for gif images.
+THe argument `imageQuality` only works for jpeg and webp images.
```
Comment (ditman) [inline packages/image_picker/image_picker_for_web/README.md:47]:
```suggestion
The argument `imageQuality` only works for jpeg and webp images.
```

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart [hunk 15]
```diff
@@ -157,8 +163,11 @@ class ImagePickerPlugin extends ImagePickerPlatform {
     double? maxWidth,
     double? maxHeight,
     int? imageQuality,
-  }) {
-    return getFiles(accept: _kAcceptImageMimeType, multiple: true);
+  }) async {
+    return await Future.wait(
+        (await getFiles(accept: _kAcceptImageMimeType, multiple: true)).map(
+            (e) => _imageResizer.resizeImageIfNeeded(
+                e, maxWidth, maxHeight, imageQuality)));
```
Comment (ditman) [inline packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:170]: Do you need the top-level await Future.wait? Isn't this enough?

```dart
return (await getFiles(...)).map( (e) => _imageResizer.resize(...) );
```

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart [hunk 16]
```diff
@@ -241,20 +250,20 @@ class ImagePickerPlugin extends ImagePickerPlatform {
   Future<List<XFile>> _getSelectedXFiles(html.FileUploadInputElement input) {
     final Completer<List<XFile>> _completer = Completer<List<XFile>>();
     // Observe the input until we can return something
-    input.onChange.first.then((event) {
+    input.onChange.first.then((event) async {
```
Comment (ditman) [inline packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:253]: There's no `await` inside this event function, so does this need to be `async`?

File: packages/image_picker/image_picker_for_web/lib/image_resizer.dart (hunk 13 above)
[the remark this answers is not in the document]
Comment (ditman) [reply packages/image_picker/image_picker_for_web/lib/image_resizer.dart:16]: > Would they need to be individually tested ?

I think that while keeping the tests that you already have for happy-case resizing, having these methods individually tested would help adding coverage for extra corner cases, yes! :)

File: packages/image_picker/image_picker_for_web/lib/image_resizer.dart [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+    final imageLoadCompleter = Completer<html.ImageElement>();
+    final imageElement = html.ImageElement();
+    imageElement.src = blobUrl;
+
+    imageElement.onLoad.listen((event) {
+      html.Url.revokeObjectUrl(blobUrl);
+      imageLoadCompleter.complete(imageElement);
+    });
+    imageElement.onError.listen((event) {
+      final excpetion = Exception(
```
Comment (ditman) [inline packages/image_picker/image_picker_for_web/lib/image_resizer.dart]:
```suggestion
      final exception = Exception(
```

(small typo)

File: packages/image_picker/image_picker_for_web/lib/image_resizer.dart [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+        file.mimeType == "image/gif") {
+      //TODO Implement maxWidth and maxHeight for image/gif
+      return file;
+    }
+    try {
+      final imageElement = await loadImage(file.path);
+      final canvas = resizeImageElement(imageElement, maxWidth, maxHeight);
+      return writeCanvasToFile(file, canvas, imageQuality);
+    } catch (e) {
+      return file;
```
Comment (ditman) [inline packages/image_picker/image_picker_for_web/lib/image_resizer.dart:28]: Note that you're revoking the blobUrl in the onLoad event of the `loadImage` method. This `return` is going to return a bad file in case `resizeImageElement` or `writeCanvasToFile` fail in any way.

I think you should `revokeObjectUrl` only after successfully calling `writeCanvasToFile` to ensure that the old blob URL is revoked after nobody else is going to use it.

File: packages/image_picker/image_picker_for_web/lib/image_resizer_utils.dart (hunk 11 above)
Comment (ditman) [reply packages/image_picker/image_picker_for_web/lib/image_resizer_utils.dart:30]: > it scales the image above its original width and height

only if `resizeFactor` is < 1, I think the upscaling can be avoided by:

```
return (resizeFactor > 1 ? imageSize ~/ resizeFactor : imageSize);
```

Right? No need for the super complex algorithm to compute the resized dimensions?

**Review by ditman — COMMENTED** [review]
Some extra comments, I think this is very close to being done!

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart (hunk 15 above)
Comment (balvinderz) [reply packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:170]: _imageResizer.resize() returns a future. so if we use this then return type is `Future<List<Future<Xfile>>>`  but we need `Future<List<XFIile>> so we need the future.wait function.

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart (hunk 16 above)
Comment (balvinderz) [reply packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:253]: done

File: packages/image_picker/image_picker_for_web/lib/image_resizer.dart (hunk 17 above)
Comment (balvinderz) [reply packages/image_picker/image_picker_for_web/lib/image_resizer.dart]: done

File: packages/image_picker/image_picker_for_web/lib/image_resizer.dart (hunk 13 above)
[the remark this answers is not in the document]
Comment (balvinderz) [reply packages/image_picker/image_picker_for_web/lib/image_resizer.dart:16]: I have added tests for each function.

File: packages/image_picker/image_picker_for_web/lib/image_resizer.dart [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+  }
+
+  /// function that loads the blobUrl into an imageElement
+  Future<html.ImageElement> loadImage(String blobUrl) {
+    final imageLoadCompleter = Completer<html.ImageElement>();
+    final imageElement = html.ImageElement();
+    imageElement.src = blobUrl;
+
+    imageElement.onLoad.listen((event) {
+      html.Url.revokeObjectUrl(blobUrl);
```
Comment (ditman) [inline packages/image_picker/image_picker_for_web/lib/image_resizer.dart:39]: This is a potential bug. If we revokeObjectUrl here, and then `resizeImageElement` or `writeCanvasToFile` fail, we're returning a `file` object (line 28) that has a revoked blob URL, so its data can't be accessed.

File: packages/image_picker/image_picker_for_web/lib/image_resizer.dart (hunk 18 above)
Comment (ditman) [reply packages/image_picker/image_picker_for_web/lib/image_resizer.dart:28]: (This is unresolved, I explained the issue in another comment)

File: packages/image_picker/image_picker_for_web/lib/image_resizer.dart [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+
+import 'dart:async';
+import 'dart:math';
+import 'dart:ui';
+import 'package:image_picker_for_web/image_resizer_utils.dart';
+import 'package:image_picker_platform_interface/image_picker_platform_interface.dart';
+import 'dart:html' as html;
+
+/// Helper class that resizes image
+class ImageResizer {
```
Comment (ditman) [inline packages/image_picker/image_picker_for_web/lib/image_resizer.dart:13]: Please, move this image_resizer and its utils to `lib/src`, so people can't `import 'package:image_picker_for_web/image_resizer.dart';` from outside of this package.

Comment (balvinderz) [reply packages/image_picker/image_picker_for_web/lib/image_resizer.dart]: done

File: packages/image_picker/image_picker_for_web/lib/image_resizer.dart (hunk 19 above)
Comment (balvinderz) [reply packages/image_picker/image_picker_for_web/lib/image_resizer.dart]: Forgot to push this change. done now.

File: packages/image_picker/image_picker_for_web/lib/image_resizer_utils.dart (hunk 11 above)
Comment (balvinderz) [reply packages/image_picker/image_picker_for_web/lib/image_resizer_utils.dart]: yes

**Comment by ditman** [comment]
@googlebot I consent.

**Comment by ditman** [comment]
I've rebased this branch with the latest master, there was a small breakage in CI that required the latest code to pass.

**Comment by ditman** [comment]
(It's too late to land this. I'll merge on monday.)

**Comment by balvinderz** [comment]
Thanks for the dash @ditman 😄
